plugins/woocommerce/src/Internal/DataStores/StockNotifications/StockNotificationsDataStore.php

```diff
@@ -0,0 +1,534 @@
+			}
+		}
+
+		$notification->save_meta_data();
+
+		if ( $changes ) {
+			$notification->apply_changes();
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Delete a stock notification.
+	 *
+	 * @param Notification $notification The data object to delete.
+	 * @param array        $args         Additional arguments.
+	 * @return void
+	 */
+	public function delete( &$notification, $args = array() ) {
+		global $wpdb;
+
+		$wpdb->delete( $this->get_table_name(), array( 'id' => $notification->get_id() ), array( '%d' ) );
+
+		$wpdb->delete( $this->get_meta_table_name(), array( 'notification_id' => $notification->get_id() ), array( '%d' ) );
+
+		$wpdb->delete( $this->get_logs_table_name(), array( 'notification_id' => $notification->get_id() ), array( '%d' ) );
+	}
+
+	/**
+	 * Add meta.
+	 *
+	 * @param Notification $notification The data object to add.
+	 * @param \stdClass    $meta         The meta object to add (containing ->key and ->value).
+	 * @return int|false The meta ID or false if the meta was not added.
+	 */
+	public function add_meta( &$notification, $meta ) {
+		$add_meta = $this->data_store_meta->add_meta( $notification, $meta );
+		$this->after_meta_change( $notification );
+		return $add_meta ? $add_meta : false;
+	}
+
+	/**
+	 * Read meta.
+	 *
+	 * @param Notification $notification The data object to read.
+	 * @return array
+	 */
+	public function read_meta( &$notification ): array {
+		$raw_meta_data = $this->data_store_meta->read_meta( $notification );
+		return $this->filter_raw_meta_data( $notification, $raw_meta_data );
+	}
+
+	/**
+	 * Update meta.
+	 *
+	 * @param Notification $notification The data object to update.
+	 * @param \stdClass    $meta         The meta object to update (containing ->id, ->key and ->value).
+	 * @return bool
+	 */
+	public function update_meta( &$notification, $meta ): bool {
+		$update_meta = $this->data_store_meta->update_meta( $notification, $meta );
+		$this->after_meta_change( $notification );
+		return $update_meta;
+	}
+
+	/**
+	 * Delete meta.
+	 *
+	 * @param Notification $notification The data object to delete.
+	 * @param \stdClass    $meta         The meta object to delete (containing at least ->id).
+	 * @return bool
+	 */
+	public function delete_meta( &$notification, $meta ): bool {
+		$delete_meta = $this->data_store_meta->delete_meta( $notification, $meta );
+
+		$this->after_meta_change( $notification );
+		return $delete_meta;
+	}
+
+	/**
+	 * Perform after meta change operations.
+	 *
+	 * @param Notification $notification The notification object.
+	 * @return bool True if changes were applied, false otherwise.
+	 */
+	private function after_meta_change( &$notification ): bool {
+
+		$current_time      = time();
+		$current_date_time = new \WC_DateTime( "@$current_time", new \DateTimeZone( 'UTC' ) );
+
+		$should_save =
+			$notification->get_id() > 0
+			&& $notification->get_date_modified( 'edit' ) < $current_date_time
+			&& empty( $notification->get_changes() );
```

Comment:
Can you expand on the logic here a bit? Do you assume that if there are any changes, someone else will call `save` on the instance of Notification?

Reply:
> Do you assume that if there are any changes, someone else will call save on the instance of Notification?

Exactly! The rationale here is:
- Changes in the notification object will effectively require a `save()`, or;
- If there are no changes, it means that meta has changed, and thus the last modified date needs updating. 

A possible drawback is that if someone seeking to update a list of props and meta starts by setting or updating the meta before another prop changes, the system will perform two saves.

It's inspired by the HPOS metadata handling [here](https://github.com/woocommerce/woocommerce/blob/13b8feed35c4f204a1d9c84860336738a05afee1/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L3338).

Reply:
Nice!